subprojects/execution/src/test/groovy/org/gradle/internal/execution/impl/steps/CacheStepTest.groovy

```diff
@@ -0,0 +1,116 @@
+    def buildCacheCommandFactory = Mock(BuildCacheCommandFactory)
+    def outputChangeListener = Mock(OutputChangeListener)
+    def currentBuildId = UniqueId.generate()
+    Step<Context, CurrentSnapshotResult> delegateStep = Mock(Step)
+    def cacheStep = new CacheStep<CachingContext>(buildCacheController, outputChangeListener, buildCacheCommandFactory, delegateStep)
+    def cacheHandler = Mock(CacheHandler)
+    def unitOfWork = Mock(UnitOfWork)
+    def loadMetadata = Mock(BuildCacheCommandFactory.LoadMetadata)
+    def cachedOriginMetadata = Mock(OriginMetadata)
+    def context = Stub(CachingContext) {
+        getCacheHandler() >> cacheHandler
+        getWork() >> unitOfWork
+    }
+
+    def "loads from cache"() {
+        def outputsFromCache = ImmutableSortedMap.of("test", new EmptyCurrentFileCollectionFingerprint())
+
+        when:
+        def result = cacheStep.execute(context)
+        def originMetadata = result.originMetadata
+        def finalOutputs = result.finalOutputs
+        then:
+        result.outcome == ExecutionOutcome.FROM_CACHE
+        result.failure == null
+        originMetadata == cachedOriginMetadata
+        finalOutputs == outputsFromCache
+
+        1 * cacheHandler.load(_) >> Optional.of(loadMetadata)
+        1 * loadMetadata.originMetadata >> cachedOriginMetadata
+        1 * loadMetadata.resultingSnapshots >> outputsFromCache
+        0 * _
+    }
+
+    def "executes work and stores in cache on cache miss"() {
+        def executionResult = new TestCurrentSnapshotResult()
+
+        when:
+        def result = cacheStep.execute(context)
+
+        then:
+        result == executionResult
+
+        1 * cacheHandler.load(_) >> Optional.empty()
+        1 * delegateStep.execute(_) >> executionResult
+        1 * cacheHandler.store(_)
+        0 * _
+    }
+
+    def "failures are not stored in the cache"() {
+        def failure = new RuntimeException("failed")
+        def failedResult = new TestCurrentSnapshotResult(failure)
+
+        when:
+        def result = cacheStep.execute(context)
+
+        then:
+        result == failedResult
+
+        1 * cacheHandler.load(_) >> Optional.empty()
+        1 * delegateStep.execute(_) >> failedResult
+        _ * unitOfWork.displayName >> "Display name"
+        0 * cacheHandler.store(_)
+        0 * _
+    }
+
+    class TestCurrentSnapshotResult implements CurrentSnapshotResult {
+        TestCurrentSnapshotResult(Throwable failure = null) {
+            this.failure = failure
+        }
+        final ImmutableSortedMap<String, CurrentFileCollectionFingerprint> finalOutputs = ImmutableSortedMap.of("test", new EmptyCurrentFileCollectionFingerprint())
+        OriginMetadata getOriginMetadata() {
+            OriginMetadata.fromCurrentBuild(currentBuildId, 0)
+        }
+        ExecutionOutcome getOutcome() {
+            ExecutionOutcome.EXECUTED
+        }
+        @Nullable
+        final Throwable failure
+    }
+}
```

Comment:
This looks a little confusing now, with some things defined as constants, others as a kind of constant, yet others passed as constructor parameters. I think it would be cleaner to implement `CurrentSnapshotResult` locally in the two places it is used.

Reply:
Implementing now in the two places where it is used.